app/views/unattended/provisioning_templates/registration/global_registration.erb

```diff
@@ -44,13 +44,19 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
 
     curl --insecure --output $CONSUMER_RPM <%= subscription_manager_configuration_url %>
     yum localinstall $CONSUMER_RPM -y
-    yum install subscription-manager -y
     subscription-manager register --org=<%= @organization.label %> --activationkey="sample-key"
-
-    rm -f $CONSUMER_RPM
 else
     create_host_curl
 fi
 <% else -%>
 create_host_curl
 <% end -%>
+
+if ! [ $? = 0 ]; then
+    echo ""
+    echo "Host registration failed"
+    exit 1
+fi
+
+TOKEN=$(curl -s -X GET "<%= foreman_server_url %>/api/hosts?search=name+%3D+$(hostname --fqdn)" <%= headers.join(' ') %> | grep -Po '(?<=\"registration_token\":")([^\"]+)')
+curl -X GET "<%= foreman_server_url %>/foreman_register/hosts/register?token=$TOKEN" | bash
```

Comment:
Why do we double `curl` here?

cc: @ekohl

Reply:
1. I need to get token for Host registration template
2. I will use the `$TOKEN` for rendering the Host registration template

Reply:
Why do we need an explicit host registration template? I think one call to Foreman for registering a host should be enough.
You could say, I'm nitpicking here, but how would you explain to a user what the global registration template does and what the host-specific template does.

Reply:
> Why do we need an explicit host registration template? I think one call to Foreman for registering a host should be enough.

Yes, that is true but we want to also do some setup on the host. In this case we want to setup insights & remote execution. 
And for this _Host registration template_ is ideal.

In _Host registration template_ we have access to the `@host` object, so users can adjust registration behavior for specific OS or host group by configuring host parameters.

For example, you want to install Insights on the RHEL & CentOS but not on Debian and others,
or setup different ssh keys for different OS.

> You could say, I'm nitpicking here, but how would you explain to a user what the global registration template does and what the host-specific template does.

Nope, you are right, the naming can be little bit confusing, but is pretty easy:
**Global Registration template** - Register unknown host to the Foreman.
**Host registration template** - Setup the host (Install puppet, insights, setup REX ...)

Reply:
Maybe it **Host setup template** would be better than **Host registration template**

Reply:
I think that calling register again is a bad design. If you want such a thing, I'd recommend implementing a special action to create the host. You currently do a `POST` to the hosts API, but what if you implement `POST /register` to create the host? That would then return a rendered template with additional actions to be performed. This avoids a request and makes authentication easier. It would also be easier on a Smart Proxy because you only need to deal with a single endpoint.

Reply:
how would you authenticate the `POST /register`? note that we get the registration token in the first curl, because we can't generate it before the host exists. And it exists only after we do the POST request.

Reply:
We spent some time on this today with @ekohl. This would work in vanilla Foreman scenario, but when subscription-manager is involved, that would be a different flow. subscription-manager is unable to execute the template it would receive and even if we implement it, it wouldn't be backported to old RHELs. We'll eliminate the need for fetching details, but we'll still need to get the `Host registration template` at the end of the `Global registration template`.

Hope this makes it easier to visualize in mind, the documentation is not available yet, but will be once the code is in... Global template is generic recipe, how to register to this Foreman instance. As part of that it preps the system (e.g. installs the cert in order to securely talk to Foreman) and creates the initial record in Foreman DB. Then it tell host to start bootstrapping. That is handled by Host registration template, a specific template for a given host. That may differ a lot based on chosen hostgroup, OS or other trivia we know about that host. That's why it's specific to the host. Getting this template switches the host into "started registration" build mode. After it's done, it call back home informing Foreman it's built (same as in case of kickstart/preseed). Thanks to that, user will know when host is ready to use.

Reply:
To be completely honest, I don't get what you're trying to say here. Maybe it's because I don't fully understand the subscription-manager based workflow, but I think we should design this in a way where it does not really matter what particular tools are used to manage the host.

Let's start with a simple question:
In the current implementation of the global registration template, why is there a basically unauthorized call to `subscription-manager`? Doesn't this belong in the host-specific registration template that has all the required information about the host, possibly a one-time activation key tailored for that system?
Can't we keep this simple and just create the host in the global registration template and handle the rest in the host-specific registration template?

Maybe having a picture visualizing the workflow would help.
Let us clarify this please, before we proceed here.

Reply:
My understanding is that there are roughly 2 ways that we support to create the Host object:
* Via API
* Having subscription-manager upload facts

Since the API is clear to you, I won't get into that. I'll also note that my understanding of subscription-manager is only based on what others have told me.

Subscription-manager is used with an activation key. That key should be treated as a secret (even though often it isn't) since it is used to authenticate the request to create a host object and issue candlepin certificates. Subman also uploads facts. It replaces the API call since it does more. Doing both would be hard and not really solve anything.

You are right that the Katello workflow is still broken because there's no macro for the exact activation key, but that depends on https://github.com/Katello/katello/pull/8903.

Reply:
How is the candlepin workflow different from creating the host based on e.g. puppet fact uploads?
I don't think supporting the subman workflow as a special workflow makes sense for the community.

Reply:
> but I think we should design this in a way where it does not really matter what particular tools are used to manage the host.

yes and that's what we're doing here, there may be some oddity in subscription-manager else branch, but I'll try to explain that too

> In the current implementation of the global registration template, why is there a basically unauthorized call to subscription-manager?

In this flow, subscription-manager is the way the host record is created. Currently it looks unauthorized, but the activation key is to be turned into the parameter. That is the authentication mechanism there. You're not the only one asking why we don't do the registration later in host registration template. The reason is, that would effectively cause re-registration because the host would already exist. Second reason is, subscription-manager sends us facts about the OS, so the host registration template can be already tailored to the OS, its version etc. For the generic flow, we don't have the common facter yet. So the best thing user can do is to set the OS on hostgroup level and specify hostgroup_id as a GRT parameter. We talked about improving this flow to either use what's available (facter || ohai || ansible -m) or start shipping uFacter (lzap's facter implemntation in go) or look at /etc/os-release and just parse the info from there. That all will take longer time than we have for this, so we want to have at least subscription-manager used as a facter for now.

> Doesn't this belong in the host-specific registration template that has all the required information about the host, possibly a one-time activation key tailored for that system?

The activation key shouldn't be loaded for the host, actually user should specify it. The current code is misleading, but we tried to split the work to more smaller PRs to move quicker.

> Can't we keep this simple and just create the host in the global registration template and handle the rest in the host-specific registration template?

We will for the non subscription-manager flow. That still requires one curl for GRT, second for HRT. With subscription-manager we get the limitation of it can't run the custom template as part of the registration. It also can't read the registration token. Therefore there will be more API calls in that branch.

Reply:
> Maybe having a picture visualizing the workflow would help.

yes, the documentation will defnitely need to contain that, once you have a moment, if you want please ping me on IRC, based on the past experience, short mtg saves thousands of lines :-)

Reply:
> How is the candlepin workflow different from creating the host based on e.g. puppet fact uploads?

With Puppet there is a server that is authenticated and authorized. The Puppet agent also sends a CSR and the Puppet CA server has a policy on that. It can be manual (default), autosign or tokens.

With subman you have the agent talking to Katello which has implemented the RHSM API. Katello then talks to Candlepin. The activation key is a static token.

The big difference is that with Puppet it can pull a catalog from the server and converge. It can use its issued client certificates to get that catalog. With subman you could do the same and implement client certificates on the HRT endpoint. Then subman wouldn't need the registration token. Not sure if it would also work with a Smart Proxy in between.

In the curl based approach you don't have any certificates so you need some authentication. The register token that was given only allows registration (and ideally would be revoked after using it once). However, it doesn't need to search for the host since the `POST` provides the token. This was suggested elsewhere as an improvement.

Reply:
Ok. Thanks for the explanation. Now I believe I at least get the reasoning. So let's talk tokens or scopes.
In the current design, how are the scopes of the tokens limited?
If I get it correctly, the GRT is retrieved with permissions the current user has. But the hrt is retrieved in an anonymous scope, is it? Can I retrieve the HRT (and secrets it contains) for a host I don't own?

A disadvantage of tokens is, that they can't easily be revoked. They're made to be used several times during a specific timespan.

Reply:
I did suggest to change the registration to `POST /register` which would return a shell script with the HRT. That would avoid the need for the second token.

Then you would only need to deal with subman. There you can use the client certificates to retrieve HRT. The complication there is if you put a Smart Proxy in between that you then need to pass the real name. However, Smart Proxy can authenticate and we can find the Smart Proxy object. If that's found, then it can trust some parameter with the real client name.

This would completely remove the need for a Host Registration Token.

Reply:
How about this flow in the GRT:

The script gathers facts about a host, in the first iteration that's just the fqdn. This can be extended by plugins. The GRT has input parameters to support other OSes, e.g. this could be a Powershell script.

(Optional/alternative part: subman register happens, ideally in the katello codebase if this requires special handling. It prefer if this were a separate PR at least)

Post call to retrieve the hrt, this passes the facts gathered in the first step. This also works when the host already exist.

Reply:
> So let's talk tokens or scopes.
In the current design, how are the scopes of the tokens limited?
If I get it correctly, the GRT is retrieved with permissions the current user has. But the hrt is retrieved in an anonymous scope, is it? Can I retrieve the HRT (and secrets it contains) for a host I don't own?

No, when you ask for GRT, JWT is rendered and put to it, it's then used to authenticate the HRT request.

> A disadvantage of tokens is, that they can't easily be revoked. They're made to be used several times during a specific timespan.

It will auto-expire in few hours.

> I did suggest to change the registration to POST /register which would return a shell script with the HRT. That would avoid the need for the second token.

It wouldn't. The `POST /register` also needs to be authenticated somehow. The advantage of this token is, it identifies the same user who rendered GRT, meaning we can still enforce the right org/loc and permissions when assigning additional objects. Personally I don't see a benefit of `POST /register` vs `POST /hosts`, the second guarantees all host attributes can be set as part of registration.

> Then you would only need to deal with subman. There you can use the client certificates to retrieve HRT. The complication there is if you put a Smart Proxy in between that you then need to pass the real name. However, Smart Proxy can authenticate and we can find the Smart Proxy object. If that's found, then it can trust some parameter with the real client name.

We can use smart proxy for authentication, but not for user identification and authorization. The request that is made through proxy forwards JWT, which identifies and authorizes the user. I don't think relying on proxy cert is the right way here.

> How about this flow in the GRT:

If I understand this correctly, you're suggesting combining host creation and HRT request into a single request, basically what Ewoud suggests with `POST /register`. IMHO we duplicate endpoints with that, we have one for host creation, second for HRT for specific host. API for host creation will always allow setting any host attribute, we'd need to make sure, register one will receive same updates. Perhaps we could mitigate that with shared apipie documentation, but still, seems odd if we already have endpoints for both. Also in future, we want to add more facts parsing using various facters. That will most likely create the host by different mean - facts upload. I assume we want to reuse facts upload endpoints there, hence it would again be separate request for host creation.

> Post call to retrieve the hrt, this passes the facts gathered in the first step. This also works when the host already exist.

I don't understand how your suggestion helps with existing hosts. Perhaps because it does not tell when exactly, the host is created.

Reply:
> I don't understand how your suggestion helps with existing hosts. Perhaps because it does not tell when exactly, the host is created.

That actually depends. In the subman use case, it's created by subman. The post call to retrieve the HRT would only create the host if it does not exist.
Note, that there might be other use cases where the host already exists. One of them could be the VM import feature we have.

Reply:
I've posted https://community.theforeman.org/t/rfc-simple-automatic-host-registration-wf/19588/33. Let's move the discussion back there. We've realized that at least 2 people involved in the RFC misunderstood what was proposed. I've added diagrams that I hope can help us avoid future confusion.